### ISSUE TITLE
iohub alternatives?

### ISSUE THREAD

**Issue by hoechenberger** [psychopy/psychopy#1541]
iohub has been sitting there pretty much unmaintained for about 1.5 years now. It provides excellent functionality, but is very complex for both developers and users. In the past we'd been thinking of switching all (input) device handling to iohub, but it just never happened, probably because no-one came up with a wrapper that would allow users to seamlessly switch their existing (non-iohub) code to iohub :)

I've been thinking about a "restart" in terms of input device handling for quite a while now, but never really found the time to actually start working on this. However, I just came across @aforren1's [toon](https://github.com/aforren1/toon), which -- among other features -- provides a simple input device handler. This handler can utilize multi-processing to run and poll devices in the background. 

Currently it supports (quoting from the `README`):

- HAND (custom force measurement device) by class Hand
- Flock of Birds (position tracker) by class BlamBirds
- Keyboard (for changes in keyboard state, more accurate timing) via Keyboard
- DebugKeyboard (for current keyboard state, constant error of up to 9 ms?) via DebugKeyboard
- Force Transducers (predecessor to HAND) by class ForceTransducers (Windows only.)

The code quality seems to be good, there are unit tests, and the code base of the main application and individual device drivers is MUCH slimmer and legible than iohub's.

I would therefore suggest we look into this package to evaluate whether it would make sense to rely on it for PsychoPy's device interaction.

**Comment by peircej** [comment]
Another option is something around Lab Streaming Layer, supporting more devices and remote connection
https://github.com/sccn/labstreaminglayer

My concern is that neither would end up *replacing* iohub because of lack of eyetracker support. So we end up with multiple half-complete solutions.

**Comment by hoechenberger** [comment]
> My concern is that neither would end up replacing iohub because of lack of eyetracker support. So we end up with multiple half-complete solutions.

LSL seems to be supporting a variety of eye trackers:
https://github.com/sccn/labstreaminglayer/wiki/SupportedDevices.wiki#supported-eye-tracking-hardware

**Comment by hoechenberger** [comment]
I'm just worried LSL may introduce a significant amount of boilerplate and additional steps for the user. In all their examples I've read so far, you are required to start an additional application on the stimulation computer that would then run in the background

**Comment by hoechenberger** [comment]
Also, I suspect LSL wouldn't solve the problem of polling devices "in the background" – we'd still have to use a separate process for this.

**Comment by aforren1** [comment]
Thanks for taking a look at my code! There are a few structural (e.g. returning heterogeneous data from the same device, gracefully managing failure on the remote process, handling mouse position properly) and administrative (e.g. documentation & examples are spotty, tests could certainly be better) issues I haven't dealt with yet.

 In principle, I don't think it would be *awful* to add support for remote connections via `multiprocessing,managers`(??), but it's something I haven't tried out. Adding eye trackers shouldn't be too bad either, though I may not have access to the proper hardware for testing.

In summary, I think it'll be at least a few weeks before `toon` is camera-ready, but I'm game for adding remote functionality and new devices. Also, I'll write up a vignette on how to implement new devices/how the guts of the base class work.

**Comment by hoechenberger** [comment]
@aforren1 CooI you're motivated to help :) – Now, I don't mean to break your motivation, but we first need to carefully evaluate different approaches here: ´toon´ may be just _one_ possible option. The issues with eye trackers I don't fully understand – what makes these devices so special? Don't they simply send their data via a network connection? Should be piece of cake to asynchronously fetch and process the data then, no? I've done networking in Python before and thought it was pretty easy.

**Comment by peircej** [comment]
Ah! I hadn't picked up that this was @aforren1's project. Got it now. ;-)

FYI I'm not actually bothered about remote monitoring (I believe in adding things when people need them rather than creating features that people in the future might want). And I don't know about the timing quality of LSL, which is much more important. 

Eye-tracking functionality is a big deal though. The other thing that ioHub achieved (that I never personally used) was that it could stream data to a file (HDF5) at high rates, and this data stream included the events for all the different hardware devices. 

Anyway, I think this could turn into a mega-thread and maybe we should get together for a hangout instead sometime?

**Comment by hoechenberger** [comment]
> Ah! I hadn't picked up that this was @aforren1's project. Got it now. ;-)
> 
> FYI I'm not actually bothered about remote monitoring (I believe in adding things when people need them rather than creating features that people in the future might want). And I don't know about the timing quality of LSL, which is much more important.

As far as I understand, everything is "remote" to LSL: It's based on a client-server architecture and uses network communication regardless of whether it's running locally or on a remote machine. For timing purposes, it uses sub-millisecond (?) time stamps to provide exact information on events.

> Eye-tracking functionality is a big deal though. The other thing that ioHub achieved (that I never personally used) was that it could stream data to a file (HDF5) at high rates, and this data stream included the events for all the different hardware devices.

Yep. Anyone know how much data a common eye tracker typically generates, as in bytes/sec?

> Anyway, I think this could turn into a mega-thread and maybe we should get together for a hangout instead sometime?

Absolutely. btw I'd also like to explore new (additional?) ways of communication -- what about gitter? Or Slack? (I grew to LOVE slack for work-related communication!) Of course, nothing can replace facte-2-face meetings & video chat, but as a complementary channel, a chat may be very beneficial from time to time!

**Comment by Schubisu** [comment]
Hi all!
Just noticed that new new issue I created [#1576](https://github.com/psychopy/psychopy/issues/1576) may also be related to [#1548](https://github.com/psychopy/psychopy/issues/1548).
Anyway, I would love to help with this issue, would it be possible to grant me access to the gitter room, if that is actively used?

Otherwise, I would suggest IRC communication (rather than slack [for some reasons](https://drewdevault.com/2015/11/01/Please-stop-using-slack.html)), as it is dead simple, fast and open. Actually I'm currently online in a #psychopy channel on the Freenode network.

**Comment by peircej** [comment]
We'll discuss the specific issue of pytables in #1548 and I've closed #1576 

I think I'm leaning towards gitter for communication (because we can use references like above to issues and commits). I've used IRC a few times and it always has a setup hurdle (servers passwords, learning the commands) so I'd prefer not to use that if possible

**Comment by peircej** [comment]
By the way, there's another alternative that I'm interested in whereby we hire Mario to port his psychtoolbox mex functions into standard dll/dylib files and then we access that from python. He's done so much work optimizing performance of lots of things, including hardware and response devices, that it really seems like the best option if we can get it going that way. (Currently talking with him about feasibility)

**Comment by m-macaskill** [comment]
Hi, I'm very late to this party, but yes we shouldn't ignore the other things that ioHub provides. It was originally recreated outside the PsychoPy project, to create a device-agnostic platform for both real-time monitoring of eye tracking data but also importantly storing data, as Jon notes. This was used in a substantial project to record many subjects on many eye trackers, to allow the creation of agreed measures of data quality, without being confounded by proprietary data formats and filtering.  It also includes common analysis procedures. I need to get back in touch with that committee to see if anyone else is still using it outside of the PsychoPy  community, and what support might be available to support its continued development and maintenance.  Meanwhile PsychoPy really does still need a simple and reliable way of collecting keypresses asynchronously, in solving that and other hardware needs, we shouldn't forgo what has already been achieved with ioHub, which offers a lot more besides.

In terms of data rates, the fastest digital trackers could be streaming at 2 kHz with, say, 100 byte packets, so could be 200 kB/s... keeping up with that is not trivial especially when considering disk i/o.

**Comment by mdcutone** [comment]
PTB and iohub are available now and are being maintained.
